Add a splash screen module

Knyazev-AI opened the pull request:
1. Реализованы элементы модуля splash.
2. Реализован сервис обработки изображения логотипа, расположенного на экране splash.
3. Реализован переход на экран авторизации после завершения показа экрана splash.
4. Удален импорт библиотеки UIKit из файлов, использующих библиотеку GoogleSignIn, по причине наличия у них аналогичного функционала.
5. Добавлен файл - изображение логотипа приложения.

Commit:
Add: logo image

Commit:
Add: image processing service

Commit:
Implement: elements of the splash module

Commit:
Remove: UIKit imports in files that use the GoogleSignIn library

The background color has also been changed

Commit:
Implement: a transition to the authorization screen after the splash screen

Curshet commented on MyBook/Modules/Application/Builder/AppBuilder.swift:
```diff
@@ -1,4 +1,3 @@
-import Foundation
```
Почему удалён импорт Foundation? Этот фреймворк по умолчанию минимально необходим в каждом файле, даже если IDE этого не требует. Необходимо вернуть импорт Foundation в файле.

Curshet commented on MyBook/Modules/Application/Coordinator/AppCoordinator.swift:
```diff
@@ -49,6 +49,20 @@ private extension AppCoordinator {
+        window.makeKeyAndVisible()
+        routers[AppCoordinatorKey.splash] = router
+        
+        DispatchQueue.main.asyncAfter(deadline: .now() + 3) { [weak self] in
```
**1.** ТЗ не предусматривало подобных задержек. Если это всё же требуются, то реализация должна быть имплеменирована внутри Splash View Model. В логике координатора такого быть не должно. Необходимо переместить данную логику. И достаточно ограничить число 2-мя секундами. В продуктовых проектах на этапе загрузки Splash могут подгружаться какие-то параметры с сервера, тогда запуск приложения может длиться и 3+ секунд, но в нашем случае в такой длинной задержке нет необходимости.

**2.** Сигнал на запуск должен исходить из VM -> Router -> AppCoordinator, а не от от главной очереди DispatchQueue.main. В случае с очередью теряется весь контроль над запуском приложения, ломается логика.

Curshet commented on MyBook/Modules/Application/Delegate/AppDelegate.swift:
```diff
@@ -1,4 +1,3 @@
-import UIKit
```
`import Foundation`

Curshet commented on MyBook/Modules/Application/Interactor/AppInteractor.swift:
```diff
@@ -1,4 +1,3 @@
-import Foundation
```
`import Foundation`

Curshet commented on MyBook/Modules/Application/Interactor/Services/AppUserLogin/AppUserLogin.swift:
```diff
@@ -1,4 +1,3 @@
-import Foundation
```
`import Foundation`

Curshet commented on MyBook/Modules/Application/Interactor/Services/AppUserLogin/AppUserLoginProtocol.swift:
```diff
@@ -1,4 +1,3 @@
-import Foundation
```
`import Foundation`

Curshet commented on MyBook/Modules/Authorization/Builder/AuthorizationBuilder.swift:
```diff
@@ -1,4 +1,3 @@
-import Foundation
```
`import Foundation`

Curshet commented on MyBook/Modules/Authorization/View/AuthorizationView.swift:
```diff
@@ -1,4 +1,3 @@
-import UIKit
```
`import Foundation`

Curshet commented on MyBook/Modules/Authorization/ViewModel/Services/AuthorizationUserLogin/AuthorizationUserLogin.swift:
```diff
@@ -1,4 +1,3 @@
-import UIKit
```
`import Foundation`

Curshet commented on MyBook/Resources/Assets.xcassets/logo.imageset/Contents.json:
Assets должен содержать те же каталоги (и подкаталоги), что содержат модули приложения. Необходимо создать каталоги: Splash, Authorization, Menu. В каталог Splash переместить изображение "logo". Имена каталогов и файлов начинаются с заглавной буквы и далее - camelCase, без пробелов. Актуально для всего проекта.

Curshet commented on MyBook/Modules/Application/Coordinator/AppCoordinator.swift:
```diff
@@ -49,6 +49,20 @@ private extension AppCoordinator {
+    
     func startAuthorizationModule(_ window: UIWindow?) {
         guard let window, let builder = builder.authorizationBuilder, let controller = builder.controller, let router = builder.router else {
             fatalError("There aren't any significant authorization module objects")
```
Метод "makeKeyAndVisible" у window вызывается в методе "startSplashModule", не нужно далее его использовать в методе "startAuthorizationModule".

Curshet commented on MyBook/Modules/Application/Coordinator/AppCoordinator.swift:
```diff
@@ -49,6 +49,20 @@ private extension AppCoordinator {
+        routers[AppCoordinatorKey.splash] = router
+        
+        DispatchQueue.main.asyncAfter(deadline: .now() + 3) { [weak self] in
+            self?.startAuthorizationModule(window)
```
Вызывается метод старта модуля авторизации, при этом роутер Splash из памяти не удаляется. Почему?

Curshet commented on MyBook/Modules/Splash/View/SplashView.swift:
```diff
@@ -1,3 +1,48 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    func sendEvent(_ image: UIImage) {
```
Данный метод является протокольным. Его необходимо выносить в отдельный extension c // MARK: Protocol.

Curshet commented on MyBook/Modules/Splash/View/SplashView.swift:
```diff
@@ -1,3 +1,48 @@
+    
+}
+
+private extension SplashView {
```
Перед любым extension необходимо писать // MARK: Private, Public, Protocol...

Curshet commented on MyBook/Modules/Splash/View/SplashView.swift:
```diff
@@ -1,3 +1,48 @@
+
+private extension SplashView {
+    
+    func setupSplashScreenView() {
```
Метод можно именовать короче: setupConfiguration. Имя класса использовать незачем, и так понятно что именно будет настраиваться. Но то, что касается subviews - иначе. К примеру, то, что касается настройки imageView логотипа в данном случае можно было бы назвать setupLogoImageView и т.д.

Curshet commented on MyBook/Modules/Splash/View/SplashView.swift:
```diff
@@ -1,3 +1,48 @@
+        setupImageViewConstraints()
+    }
+    
+    func setupImageViewConstraints() {
```
Так как в данном случае у нас всего 1 subview, то метод вовсе можно назвать setupConstraints.

Curshet commented on MyBook/Modules/Splash/View/SplashView.swift:
```diff
@@ -1,3 +1,48 @@
+class SplashView: UIView, SplashViewProtocol {
+    
+    var action: (() -> Void)?
+    private let imageView: UIImageView
```
Плохой нейминг. ImageView это название компонента UIKit, человеку не понятно что это именно такое как свойство объекта. Вариантом получше будет назвать это поле как "logoImageView".

Curshet commented on MyBook/Modules/Splash/ViewController/SplashViewController.swift:
```diff
@@ -9,10 +9,26 @@ class SplashViewController: UIViewController {
+            self?.myView.sendEvent(image)
+        }
+        
+        myView.action = { [weak self] in
```
В текущей задаче внешних событий от view не предполагается вовсе. Реализованный наблюдатель у view явно лишний. Необходимо его убрать. Запрос данных у VM должен происходить в соотв. методах жизненного цикла VC.

Curshet commented on MyBook/Modules/Splash/ViewController/SplashViewController.swift:
```diff
@@ -9,10 +9,26 @@ class SplashViewController: UIViewController {
+    
+    private func setupObservers() {
+        viewModel.action = { [weak self] in
+            guard let image = UIImage(data: $0) else { return }
```
Данной проверки на опциональность здесь быть не должно. Необходимо выполнять её либо в VM либо сделать аргумент типа UIImage у view опциональным и избавиться от проверки вовсе.

Curshet commented on MyBook/Modules/Splash/ViewModel/Services/SplashLogoImage/SplashLogoImage.swift:
Не понятно зачем ради одного UIImage создан целый сервис, который это изображение отправляет? При запросе данных из VM достаточно сразу же отправлять <UIImage(named: "logo")> без доп. сервиса. Конвертация в "pngData" не требуется. От сервиса необходимо избавиться.

Curshet reviewed (commented):
**1.** Создать каталоги в "Assets" и поправить нейминг файлов.
**2.** Исправить архитектурные замечания.
**3.** Приложение в начале стартует с белого экрана и далее появляется логотип. Поправить старт таким образом, чтобы логотип появлялся сразу при запуске.
**4.** Отсутствует лейбл статуса PR.
**5.** Именование ветки не очень информативное. Хороший вариант именования: **feature/ios-add-splash-screen-module**. Название ветки содержит информацию о платформе и краткое информативное описание работы. Сейчас исправлять не нужно, на будущее.
**6.** В именовании PR не нужно явно указывать: "...to the project". Это очевидная вещь, что добавление кодовой базы происходит в проект. Хороший вариант именования PR: **Add a splash screen module**. Именование PR обычно содержит в себе именование ветки и/или дополняет его.
**7.** Тег коммита "Implement" по сути означает то же самое, что и "Add". Думаю, что достаточно использовать только последний.

Commit:
Merge branch 'develop' into feature/ios-add-splash

Commit:
Fix: curshet's remarks on PR #4 "Add a Splash screen module"

Including:
1. Importing the Foundation library to the files specified in the comments;
2. Create directories in the Resources section and configure file naming;
3. Removing the logo image processing service on the splash screen;
4. Elimination of architectural remarks;
5. The implementation of the start of the application with the logo and the corresponding background color of the project;
6. Resolving conflicts related to the infusion of the develop branch, where new project solutions are implemented

Curshet commented on MyBook/Modules/Application/Coordinator/AppCoordinator.swift:
```diff
@@ -49,20 +49,35 @@ private extension AppCoordinator {
+        }
+        
+        window.rootViewController = controller
+        window.makeKeyAndVisible()
```
Вызов "makeKeyAndVisible" дублируется в AppBuilder и данном методе.

Commit:
Delete: calling window.makeKeyAndVisible() from the AppCoordinator file
